wp-modules/pattern-data-handlers/pattern-data-handlers.php

```diff
@@ -363,15 +365,56 @@ function construct_pattern_php_file_contents( $pattern_data ) {
  * Viewport Width: ' . $pattern['viewportWidth'] . '
  * Block Types: ' . implode( ', ', $pattern['blockTypes'] ) . '
  * Post Types: ' . implode( ', ', $pattern['postTypes'] ) . '
- * Inserter: ' . ( $pattern['inserter'] ? 'true' : 'false' ) . '
- */
-
+ * Inserter: ' . ( $pattern['inserter'] ? 'true' : 'false' ) . maybe_add_custom_category_header( $pattern['customCategories'] ) . '
+ */' . create_formatted_category_registrations( $pattern['customCategories'] ) . '
 ?>
 ' . trim( $pattern['content'] ) . '
 ';
 	return $file_contents;
 }
 
+/**
+ * Returns a string that conditionally contains the custom category header.
+ *
+ * @param array $custom_categories The custom category titles/labels.
+ * @return string
+ */
+function maybe_add_custom_category_header( $custom_categories ) {
+	$custom_category_header = '';
+
+	if ( ! empty( $custom_categories ) ) {
+		$custom_category_header = "\n * Custom Categories: " . implode( ', ', $custom_categories );
+	}
+
+	return $custom_category_header;
+}
+
+/**
+ * Returns a formatted string that will register custom categories from the pattern file.
+ *
+ * @param array $custom_categories The custom category titles/labels to be parsed.
+ * @return string
+ */
+function create_formatted_category_registrations( $custom_categories ) {
+	$custom_category_registrations = '';
+
+	if ( ! empty( $custom_categories ) ) {
+		$custom_category_registrations = array_map(
+			function ( $category_label ) {
+				$category_name = strtolower( str_replace( ' ', '-', $category_label ) );
+				$text_domain   = wp_get_theme()->get( 'TextDomain' );
+				$label_arr     = $text_domain ? "[ 'label' => __( '$category_label', '$text_domain' ), 'pm_custom' => true ]" : "[ 'label' => '$category_label', , 'pm_custom' => true ]";
+				return "register_block_pattern_category( '$category_name', $label_arr );";
+			},
+			$custom_categories,
+		);
+
+		$custom_category_registrations = "\n" . implode( "\n", $custom_category_registrations );
+	}
+
+	return $custom_category_registrations;
+}
+
```

Comment:
In a future PR, let's split this file into another. It's getting long.

Reply:
Agreed — we could clean up this file dramatically by splitting things up.